codegen/src/api/mod.rs

```diff
@@ -376,40 +438,41 @@ pub fn generate_structs_from_variants<F>(
     variant_to_struct_name: F,
     error_message_type_name: &str,
     crate_path: &CratePath,
-) -> Vec<(String, CompositeDef)>
+) -> Result<Vec<(String, CompositeDef)>, CodegenError>
 where
     F: Fn(&str) -> std::borrow::Cow<str>,
 {
     let ty = type_gen.resolve_type(type_id);
-    if let scale_info::TypeDef::Variant(variant) = ty.type_def() {
-        variant
-            .variants()
-            .iter()
-            .map(|var| {
-                let struct_name = variant_to_struct_name(var.name());
-                let fields = CompositeDefFields::from_scale_info_fields(
-                    struct_name.as_ref(),
-                    var.fields(),
-                    &[],
-                    type_gen,
-                );
-                let struct_def = CompositeDef::struct_def(
-                    &ty,
-                    struct_name.as_ref(),
-                    Default::default(),
-                    fields,
-                    Some(parse_quote!(pub)),
-                    type_gen,
-                    var.docs(),
-                    crate_path,
-                );
-                (var.name().to_string(), struct_def)
-            })
-            .collect()
-    } else {
-        abort_call_site!(
-            "{} type should be an variant/enum type",
-            error_message_type_name
-        )
+
+    let scale_info::TypeDef::Variant(variant) = ty.type_def() else {
+        return Err(CodegenError::InvalidType(error_message_type_name.into()))
+    };
+
+    let mut result = Vec::new();
+
+    for var in variant.variants() {
```

Comment:
You could collect this this a `Result<Vec<_>, Error>` instead to avoid creating the vec and push.

Readable as it is right now, so optional :)